### PR TITLE
topdown/every: improve traces

### PR DESCRIPTION
**Pull request by srenatus**
To test this, I've used calls like `go test ./topdown -run TestRego/every/simple_failure -v`, their output now includes the traces if `-v` is passed:

```
=== RUN   TestRego
=== RUN   TestRego/every/simple_failure,_first
Enter data.test.p = x
| Eval data.test.p = x
| Index data.test.p (matched 1 rule, early exit)
| Enter data.test.p
| | Eval __local2__ = [1, 2]
| | Eval every v in __local2__ { neq(v, 1) }
| | Enter every v in __local2__ { neq(v, 1) }
| | | Eval __local2__[__local0__] = v
| | | Enter neq(v, 1)
| | | | Eval neq(v, 1)
| | | | Fail neq(v, 1)
| | | Redo every v in __local2__ { neq(v, 1) }
| | | Redo __local2__[__local0__] = v
| | | Redo __local2__[__local0__] = v
| | | Fail every v in __local2__ { neq(v, 1) }
| | Fail every v in __local2__ { neq(v, 1) }
| | Redo __local2__ = [1, 2]
| Fail data.test.p = x
=== RUN   TestRego/every/simple_failure,_last
Enter data.test.p = x
| Eval data.test.p = x
| Index data.test.p (matched 1 rule, early exit)
| Enter data.test.p
| | Eval __local2__ = [1, 2]
| | Eval every v in __local2__ { neq(v, 2) }
| | Enter every v in __local2__ { neq(v, 2) }
| | | Eval __local2__[__local0__] = v
| | | Enter neq(v, 2)
| | | | Eval neq(v, 2)
| | | | Exit neq(v, 2) early
| | | Redo neq(v, 2)
| | | | Redo neq(v, 2)
| | | Redo every v in __local2__ { neq(v, 2) }
| | | Redo __local2__[__local0__] = v
| | | Enter neq(v, 2)
| | | | Eval neq(v, 2)
| | | | Fail neq(v, 2)
| | | Redo every v in __local2__ { neq(v, 2) }
| | | Redo __local2__[__local0__] = v
| | | Fail every v in __local2__ { neq(v, 2) }
| | Fail every v in __local2__ { neq(v, 2) }
| | Redo __local2__ = [1, 2]
| Fail data.test.p = x
--- PASS: TestRego (0.22s)
    --- PASS: TestRego/every/simple_failure,_first (0.00s)
    --- PASS: TestRego/every/simple_failure,_last (0.00s)
```

--------

Another piece to the #2090 puzzle.

### CODE REVIEW

File: ast/compile_test.go [hunk 1]
```diff
@@ -2976,7 +2976,7 @@ func TestRewriteDeclaredVars(t *testing.T) {
 				p = true {
 					__local0__ = 10
 					__local3__ = [1]
-					every __local1__, __local2__ in __local3__ { __local2__ == 1 }
+					every __local1__, __local2__ in __local3__ { __local2__ = 1 }
```
Comment (johanfylling) [inline ast/compile_test.go:2979]: I was confused by the lone `__local1__`, but when I saw in the trace that it's used for lookup in `__local3__`, it makes sense. I think .. This is stuff I haven't looked at until now 😅 .

**Review by johanfylling — COMMENTED** [review]
This is touching stuff that is still outside my comfort zone, so I don't dare pass judgement on the solution as a whole. Code looks good though 😊.

**Review by tsandall — COMMENTED** [review]
Two thoughts...

1. I wonder if we should add some extra functions for tracing `every` expressions so that we can make the output a bit more meaningful. E.g., if we did `traceRedo(every)` instead of `traceRedo(every.Domain)` then in theory we could annotate the trace event with something to make it clear that the event relates to the domain. We could also hide generated variables that a user will have no context for. E.g., currently, given `every x in y {...}` we'll generate a variable for the key position of the `every` expression and it'll show up in the trace...  if the tracer knew the event was for the domain it could hide the generated variable entirely.

2. Because we're calling directly into `eval()` for the body (e.g., `body.eval(...)`), we don't get enter/exit statements for the body like I was expecting. Is there a reason for that. What I was expecting to see from the trace was an exit for each enter/redo on the body if all elements in the domain are satisfied. If any of the elements in the domain do not satisfy the body, then I would expect at least one fail from body. In terms of analysis, if OPA were to short-circuit on fail in this case, we could easily report the offending element and expression which would be useful for explanations.

**Comment by srenatus** [comment]
❓ We could take this further by replacing vars in events with values, i.e. use `event.Locals`. But I think this would diverge from what we've done so far, and it may even be confusing: it's nice to have `Enter every k, v in [1, 2]`, but the `Eval __local2__ = [1, 2]` and `Enter __local2__[__local0__] = __local1__` events would make pretty little sense then. 🤔 

So I think this might be a good place to not take things further...?

**Comment by tsandall** [comment]
> ❓ We could take this further by replacing vars in events with values, i.e. use event.Locals. But I think this would diverge from what we've done so far, and it may even be confusing: it's nice to have Enter every k, v in [1, 2], but the Eval __local2__ = [1, 2] and Enter __local2__[__local0__] = __local1__ events would make pretty little sense then. 🤔

> So I think this might be a good place to not take things further...?

Agreed, that replacing vars w/ values should be handled separately. We've tried that in the past and the problem always boils down to the data--if you just display the entire value naively then the output is too noisy. I could imagine doing something to make it clear that the child queries under the Enter are for the domain and the body respectively, but I don't think we need to block on that.

I pulled down the changes and noticed one thing: I expected the Redo events on the `every` expression to be presented as the same indentation/depth as the Enter... however, because of the current formatting logic, that's not the case. This isn't a huge deal--the query IDs are set correctly so it's really just a presentation thing. If you compare the Enter/Redo indentation for rules and comprehensions versus every, you will see what I mean...

**trace for every**

```
> every x in [1,2] { x > 0; x < 100 }
query:1     Enter __localq2__ = [1, 2]; every __localq0__, x in __localq2__ { gt(x, 0); lt(x, 100) }
query:1     | Eval __localq2__ = [1, 2]
query:1     | Eval every x in __localq2__ { gt(x, 0); lt(x, 100) }
query:1     | Enter every x in __localq2__ { gt(x, 0); lt(x, 100) }
query:1     | | Eval __localq2__[__localq0__] = x
query:1     | | Enter gt(x, 0); lt(x, 100)
query:1     | | | Eval gt(x, 0)
query:1     | | | Eval lt(x, 100)
query:1     | | | Exit gt(x, 0); lt(x, 100) early
query:1     | | Redo gt(x, 0); lt(x, 100)
query:1     | | | Redo lt(x, 100)
query:1     | | | Redo gt(x, 0)
query:1     | | Redo every x in __localq2__ { gt(x, 0); lt(x, 100) }           <--- THIS WAS SURPRISING TO ME
query:1     | | Redo __localq2__[__localq0__] = x
query:1     | | Enter gt(x, 0); lt(x, 100)
query:1     | | | Eval gt(x, 0)
query:1     | | | Eval lt(x, 100)
query:1     | | | Exit gt(x, 0); lt(x, 100) early
query:1     | | Redo gt(x, 0); lt(x, 100)
query:1     | | | Redo lt(x, 100)
query:1     | | | Redo gt(x, 0)
query:1     | | Redo every x in __localq2__ { gt(x, 0); lt(x, 100) }
query:1     | | Redo __localq2__[__localq0__] = x
query:1     | Exit __localq2__ = [1, 2]; every __localq0__, x in __localq2__ { gt(x, 0); lt(x, 100) }
query:1     Redo __localq2__ = [1, 2]; every __localq0__, x in __localq2__ { gt(x, 0); lt(x, 100) }
query:1     | Redo every x in __localq2__ { gt(x, 0); lt(x, 100) }
query:1     | | Exit every x in __localq2__ { gt(x, 0); lt(x, 100) }
query:1     | Redo __localq2__ = [1, 2]
true
```

**trace for partial rule**
```
> p[x] { some x in [1,2]; x > 0; x < 100 }
Rule 'p' defined in package repl. Type 'show' to see rules.
> p
query:1     Enter data.repl.p = _
query:1     | Eval data.repl.p = _
query:1     | Index data.repl.p (matched 1 rule)
query:1     | Enter data.repl.p
query:1     | | Eval __local3__ = [1, 2]
query:1     | | Eval x = __local3__[__local0__]
query:1     | | Eval gt(x, 0)
query:1     | | Eval lt(x, 100)
query:1     | | Exit data.repl.p
query:1     | Redo data.repl.p
query:1     | | Redo lt(x, 100)
query:1     | | Redo gt(x, 0)
query:1     | | Redo x = __local3__[__local0__]
query:1     | | Eval gt(x, 0)
query:1     | | Eval lt(x, 100)
query:1     | | Exit data.repl.p
query:1     | Redo data.repl.p
query:1     | | Redo lt(x, 100)
query:1     | | Redo gt(x, 0)
query:1     | | Redo x = __local3__[__local0__]
query:1     | | Redo __local3__ = [1, 2]
query:1     | Exit data.repl.p = _
query:1     Redo data.repl.p = _
query:1     | Redo data.repl.p = _
[
  1,
  2
]
```

**trace for complete rule**

```
> q { some x in [1,2]; x > 0; x < 100 }
Rule 'q' defined in package repl. Type 'show' to see rules.
> q
query:1     Enter data.repl.q = _
query:1     | Eval data.repl.q = _
query:1     | Index data.repl.q (matched 1 rule, early exit)
query:1     | Enter data.repl.q
query:1     | | Eval __local7__ = [1, 2]
query:1     | | Eval x = __local7__[__local3__]
query:1     | | Eval gt(x, 0)
query:1     | | Eval lt(x, 100)
query:1     | | Exit data.repl.q early
query:1     | Exit data.repl.q = _
query:1     Redo data.repl.q = _
query:1     | Redo data.repl.q = _
query:1     | Redo data.repl.q
query:1     | | Redo lt(x, 100)
query:1     | | Redo gt(x, 0)
query:1     | | Redo x = __local7__[__local3__]
query:1     | | Redo __local7__ = [1, 2]
true

```

**trace for comprehension**
```
> {x | some x in [1,2]; x > 0; x < 100 }
query:1     Enter {x | __localq0__ = [1, 2]; x = __localq0__[__localq0__]; gt(x, 0); lt(x, 100)} = _
query:1     | Eval {x | __localq0__ = [1, 2]; x = __localq0__[__localq0__]; gt(x, 0); lt(x, 100)} = _
query:1     | Enter __localq0__ = [1, 2]; x = __localq0__[__localq0__]; gt(x, 0); lt(x, 100)
query:1     | | Eval __localq0__ = [1, 2]
query:1     | | Eval x = __localq0__[__localq0__]
query:1     | | Eval gt(x, 0)
query:1     | | Eval lt(x, 100)
query:1     | | Exit __localq0__ = [1, 2]; x = __localq0__[__localq0__]; gt(x, 0); lt(x, 100)
query:1     | Redo __localq0__ = [1, 2]; x = __localq0__[__localq0__]; gt(x, 0); lt(x, 100)
query:1     | | Redo lt(x, 100)
query:1     | | Redo gt(x, 0)
query:1     | | Redo x = __localq0__[__localq0__]
query:1     | | Eval gt(x, 0)
query:1     | | Eval lt(x, 100)
query:1     | | Exit __localq0__ = [1, 2]; x = __localq0__[__localq0__]; gt(x, 0); lt(x, 100)
query:1     | Redo __localq0__ = [1, 2]; x = __localq0__[__localq0__]; gt(x, 0); lt(x, 100)
query:1     | | Redo lt(x, 100)
query:1     | | Redo gt(x, 0)
query:1     | | Redo x = __localq0__[__localq0__]
query:1     | | Redo __localq0__ = [1, 2]
query:1     | Exit {x | __localq0__ = [1, 2]; x = __localq0__[__localq0__]; gt(x, 0); lt(x, 100)} = _
query:1     Redo {x | __localq0__ = [1, 2]; x = __localq0__[__localq0__]; gt(x, 0); lt(x, 100)} = _
query:1     | Redo {x | __localq0__ = [1, 2]; x = __localq0__[__localq0__]; gt(x, 0); lt(x, 100)} = _
[
  1,
  2
]
```

**Comment by srenatus** [comment]
@tsandall I had briefly believed it was easy to fix but it's not (see revert). Let's keep it as-is for now, I suppose.

**Pull request merged**